shared/models/notification.rb

```diff
@@ -0,0 +1,70 @@
+require "securerandom"
+
+module FastlaneCI
+  # All metadata about a notification.
+  class Notification
+    # Is a UDID so we're not open to ID guessing attacks
+    #
+    # @return [String]
+    attr_accessor :id
+
+    # The relative priority of a notification
+    #
+    # @return [String]
+    attr_accessor :priority
+
+    # The name of a notification, which combined with the message can uniquely
+    # specify it
+    #
+    # @return [String]
+    attr_accessor :name
+
+    # Notification message to be displayed in the dashboard
+    #
+    # @return [String]
+    attr_accessor :message
+
+    # The time the notification was created
+    #
+    # @return [String]
+    attr_accessor :created_at
+
+    # The last time the notification was updated
+    #
+    # @return [String]
+    attr_accessor :updated_at
+
+    # Instantiates a new `Notification` model object
+    #
+    # @param  [String] id
+    # @param  [String] priority
+    # @param  [String] name
+    # @param  [String] message
+    # @return [nil]
+    def initialize(id: nil, priority: nil, name: nil, message: nil, created_at: nil, updated_at: nil)
+      self.id = id || SecureRandom.uuid
+      self.priority = %w[HIGH MEDIUM LOW].include?(priority) ? priority : "LOW"
+      self.name = name
+      self.message = message
+      self.created_at = created_at || Time.now.to_s
+      self.updated_at = updated_at || Time.now.to_s
+    end
+
+    # A way to uniquely specify a notification
+    #
+    # @return [String]
+    def primary_key
```

Comment:
This is interesting. Instead of `make_primary_key` maybe we need to define a contract for notifications. As in, if you wish to send a notification that should be deduplicated, you need to figure out how to scope it so that we can apply deduplication logic to it.

I say this because I'm scared that we're hiding the logic of how a primary key works behind the `name` variable. If you create two notifications named the same thing, they are seen as the same, but they might not be.

Instead, I'd rather register my own primary-key generator with the notification system for each module I build, `iff` I need deduplication.

Also, I think notifications should have types.
`acknowledgement_required` and `disappearing` (like a Toast in android)

And then, because of this they should also have `acknowledged` state, so that if we have deduplication logic, it will only dedupe records that are still `acknowledged? == false`, that way if we do have a reoccurring notification, it will continue to notify when the user dismisses the notification.

Does this make sense?

Also, I think it would really help to make `use cases` around how we expect to use notifications so we can explore the situations we want to support 😄

Reply:
BTW, I'm happy to work on those use cases, it might be good experience for you to flesh them out, just let me know  what you want to do 🚀